src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+import abc
+import pathlib
+import types
+from typing import Any, Callable, Dict, Literal, Mapping, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+
+import optax
+from flax.training import train_state
+
+from ott.geometry import costs, pointcloud
+from ott.problems.linear import linear_problem
+from ott.problems.quadratic import quadratic_problem
+from ott.solvers.linear import sinkhorn
+
+__all__ = ["BaseNeuralSolver", "ResampleMixin", "UnbalancednessMixin"]
+
+
+class BaseNeuralSolver(abc.ABC):
+  """Base class for neural solvers.
+
+  Args:
+    iterations: Number of iterations to train for.
+    valid_freq: Frequency at which to run validation.
+  """
+
+  def __init__(self, iterations: int, valid_freq: int, **_: Any):
+    self.iterations = iterations
+    self.valid_freq = valid_freq
+
+  @abc.abstractmethod
+  def setup(self, *args: Any, **kwargs: Any):
+    """Setup the model."""
+
+  @abc.abstractmethod
+  def __call__(self, *args: Any, **kwargs: Any):
+    """Train the model."""
+
+  @abc.abstractmethod
+  def transport(self, *args: Any, forward: bool, **kwargs: Any) -> Any:
+    """Transport."""
+
+  @abc.abstractmethod
+  def save(self, path: pathlib.Path):
+    """Save the model."""
+
+  @abc.abstractmethod
+  def load(self, path: pathlib.Path):
+    """Load the model."""
+
+  @property
+  @abc.abstractmethod
+  def training_logs(self) -> Dict[str, Any]:
+    """Return the training logs."""
+
+
+class ResampleMixin:
+  """Mixin class for mini-batch OT in neural optimal transport solvers."""
+
+  def _resample_data(
+      self,
+      key: jax.random.KeyArray,
+      tmat: jnp.ndarray,
+      source_arrays: Tuple[jnp.ndarray, ...],
+      target_arrays: Tuple[jnp.ndarray, ...],
+  ) -> Tuple[jnp.ndarray, ...]:
+    """Resample a batch according to coupling `tmat`."""
+    tmat_flattened = tmat.flatten()
+    indices = jax.random.choice(key, len(tmat_flattened), shape=[tmat.shape[0]])
+    indices_source = indices // tmat.shape[1]
+    indices_target = indices % tmat.shape[1]
+    return tuple(
+        b[indices_source, :] if b is not None else None for b in source_arrays
+    ), tuple(
+        b[indices_target, :] if b is not None else None for b in target_arrays
+    )
+
+  def _sample_conditional_indices_from_tmap(
+      self,
+      key: jax.random.PRNGKeyArray,
+      tmat: jnp.ndarray,
+      k_samples_per_x: Union[int, jnp.ndarray],
+      source_arrays: Tuple[jnp.ndarray, ...],
+      target_arrays: Tuple[jnp.ndarray, ...],
+      *,
+      source_is_balanced: bool,
+  ) -> Tuple[jnp.array, jnp.array]:
+    batch_size = tmat.shape[0]
+    left_marginals = tmat.sum(axis=1)
+    if not source_is_balanced:
+      key, key2 = jax.random.split(key, 2)
+      indices = jax.random.choice(
+          key=key2,
+          a=jnp.arange(len(left_marginals)),
+          p=left_marginals,
+          shape=(len(left_marginals),)
+      )
+    else:
+      indices = jnp.arange(batch_size)
+    tmat_adapted = tmat[indices]
```

Comment:
There's no need to run the `tmat_adapted = tmat[indices]` when the indices are just `arange`.